### ISSUE TITLE
Improve sorting in value maps

### ISSUE THREAD

**Issue by saemiw** [opengisch/QField#182]
The values offered in value maps are always sorted alphabetically by the description field. This makes sense in all cases where the field is text. In integer fields this does not always make sense. Below two examples to illustrate this (so both situations can happen in reality):

Where alphabetic sorting by description **makes sense** (left how data are entered in the value map widget on QGIS, right how the form shows value upon data entry on QGIS)
![sensfull sorting](https://user-images.githubusercontent.com/20398621/31993092-7036402c-b97c-11e7-8c83-680cb3d57467.jpg)

Where alphabetic sorting by description **makes no sense** (left how data are entered in the value map widget on QGIS, right how the form shows value upon data entry on QGIS):
![non-sense sorting](https://user-images.githubusercontent.com/20398621/31993104-77a73f64-b97c-11e7-9e70-9e35ab3fb533.jpg)

Possible solutions (needs modification on QGIS and QField):
* stop sorting the values entered in the value map widget. Use the order by which data are entered. Problem: already existing projects will behave differently upon update of QGIS.
* Add option "sort by value". By default it is unchecked. Sorting would be alphabetic for text fields and for integer or decimal values do a "numberic sorting" if the checkbox is checked. See illustration below.
![sort by value box](https://user-images.githubusercontent.com/20398621/31993113-7f1c926c-b97c-11e7-8fdb-dbacbb61f41e.jpg)

Maybe there is more options I haven't thought about. Is this doable? In which QGIS-Version? Does this need more coordination with other needs?

**QField version:** [0.10.1]

## Additional information:

* Problem started happening recently, didn't happen in an older version of QField: [No]
* Problem can be reliably reproduced, doesn't happen randomly: [Yes]
* Problem happens with all files and projects, not only some files or projects: [Yes]
* This is a "problem" of QGIS not QField. But becomes obvious when using QField.

**Comment by m-kuhn** [comment]
I think we could do that in QGIS itself as well, currently the sort order is just random. I think we can do without an extra checkbox, it should respect the order in the table (which can be adjusted by clicking on the Description or Value table headers)

**Comment by saemiw** [comment]
Are you sure it is random. I am quite sure it is alphabetically (it would be a lot of chance if my test randomly were sorted alphabetically).

> I think we can do without an extra checkbox, it should respect the order in the table (which can be adjusted by clicking on the Description or Value table headers)

Yes that would be very nice. Just to avoid misunderstandings: At the moment I cannot change the order of the table by clicking on the headers (QGIS 2.18.13 value map widget) AND QGIS reshuffels the order. So those 2 points would need your work, yes?

`

**Comment by m-kuhn** [comment]
Not sure exactly, I was just talking from memory.
At some point we had in one of our QGIS courses a widget with a rating:

 1 : Excellent
 2 : Good
 3 : Well
 4 : Come on
 5 : Better stay out

And I remember that they were ordered (in the config dialog or the widget or both) differently than expected. Maybe it was alphabetical, it just wasn't what I had expected.

> So those 2 points would need your work, yes?

I guess we could do something

**Comment by saemiw** [comment]
![prasentation1](https://user-images.githubusercontent.com/20398621/32010399-fcb10534-b9b1-11e7-8ec6-dbb5b39f9e30.jpg)

**Comment by m-kuhn** [comment]
Sidenote: this will affect the configuration interface as well and therefore be tied to an official release with which the configuration for this will be available.

**Comment by saemiw** [comment]
But it is possible within 2.18 or are there no further releases anymore for the 2. version?

**Comment by m-kuhn** [comment]
If it needs to be compatible with 2.18 some extra magic will need to be put into a plugin

**Comment by saemiw** [comment]
We were thinking of continuing using QGIS2.18 for a while but then will switch to QGIS 3. 

2 follow-up questions:
-What plugin would needs magic (qfieldsync or qgis2compat)?
- Do you think QGIS3 and QField are stable enough for us to switch already. Data loss or other major problem is not an option we can risk...

**Comment by m-kuhn** [comment]
> What plugin would need magic (qfieldsync or qgis2compat)?

qfieldsync

> Do you think QGIS3 and QField are stable enough for us to switch already? Data loss or other major problem is not an option we can risk...

Feel free to install QGIS 3.0 side-by-side with QGIS 2.18 and make your own picture and report feedback so we have a stable initial QGIS 3.0 release towards the end of this year.

The data in databases/files shouldn't be exposed to an increased risk compared to working with it with QGIS 2.
The only thing I would recommend is to work on copies of QGIS projects (because the project file structure changed).

**Comment by 3nids** [comment]
I think that keeping the elements sorted as in the table could be a fix for 2.18 while adding new sorting options would be 3.x only.
If I read correctly, @saemiw you could live with the elements sorted as in the map, correct?

**Comment by saemiw** [comment]
yes that would be a good solution for us!

**Comment by signedav** [comment]
Actually the order in the configuration-table is always sorted by the description as well after applying.
(It's because in the background is a map where "Description" is key and "Value" is value)

But of course we could display it ordered by value when we will sort the widget (table). Also the widget (combobox) in the attribute form.

Means that order will always (2.18, 3.0 and QField) be ascending by the value.
Would that be okay?

As far I understand your input, we could handle this as "fix" for 3.0 @3nids ?

**Comment by 3nids** [comment]
I think we can add an option to sort by Value or Description and ASC/DESC for 2.x and 3.0.
For 3.2+ we could add an option to introduce a custom order.

Or better:
* for 2.x and 3.0 we just force to order by value asc.
* for 3.2 we let the user order the elements as he wants in the table (by adding sorting in the table header) and we save that order. 

Does it make sense?

**Comment by signedav** [comment]
Yes. It makes more sense to sort by value than by the "Description" - at least from the users point of view.

I support the idea to force the sorting by value ascending and in to make it more configurable for the user and save the order in 3.2.

**Comment by saemiw** [comment]
A little addition to your discussion:
just keep in mind that sorting by description can make sense when you have a lot of values that have a natural order (=value field). The sorting by description can still be helpfull because everyone knows the alphabet. Example bird species:


sysnum | namedt | namefz
-- | -- | --
0 | Mischart | Oiseau hybride
0 | Art unbekannt | Espèce inconnue
20 | Gelbbrustpfeifgans | Dendrocygne fauve
30 | Javapfeifgans | Dendrocygne siffleur
40 | Schwarzschwan | Cygne noir
50 | Höckerschwan | Cygne tuberculé
60 | Zwergschwan | Cygne de Bewick
70 | Singschwan | Cygne chanteur
80 | Saatgans | Oie des moissons
81 | Saatgans (A.f.fabalis) | Oie des moissons (A.f.fabalis)

The sysnum is the systematic order (i.e. the order in which the birds are listed in every bird book). For people that know this order very well (scientists) sorting by this number is very convenient. On the other hand sorty by namedt is very convenient for normal bird watchers that don't know the systematic order but know the name of a bird they have to records. So both cases be usefull....

**Comment by 3nids** [comment]
You're right, sorry, you already pointed it quite well at the beginning.

So, I guess we should leave QGIS 2.x as is.
And keep the sorting as in the config widget.
We could still push that to 3.0 as a bugfix.

**Comment by signedav** [comment]
Just for info. At the moment it's sorted after "Description" in 2.18 and in 2.99 when storing widget data and not like it's entered:
![sorting](https://user-images.githubusercontent.com/28384354/34951113-7eba14dc-fa15-11e7-964d-3a73ab17c3bb.gif)
It does not make any kind of sorting on purpose - this is how the data are stored.

If this is kind of a "bug" we could sort it after the value. But if this would not solve the request anyway I suppose we should go for another solution. For example to save the way like it's created by the user. But I'm not sure if we still can call it a Fix then.

**Comment by m-kuhn** [comment]
store it as qlist and call it a fix was my initial intention.

**Comment by signedav** [comment]
Alright. Then I make it like this. Then it's stored and displayed like the user enters it.

**Comment by signedav** [comment]
That's the solution in QGIS 3.0. **Is it what is currently needed?**
If yes, I will make it for QGIS 2.18 and QField as well.

Sollution:
![valuemap_good-sorting](https://user-images.githubusercontent.com/28384354/34984291-b7b5ab3e-fab0-11e7-8f7f-38da2812001a.gif)

Before it was like this:
![valuemap_bad-sorting](https://user-images.githubusercontent.com/28384354/34984196-5e62d944-fab0-11e7-8bf9-06c040c7568b.gif)

**Comment by m-kuhn** [comment]
not for 2.18, it's an incompatible project file change.

**Comment by saemiw** [comment]
@,signedav if it works as depicted in the gif above under QGIS 3, then I consider our "demand" as fullfilled! 

For this years fieldwork (with Qgis 2.18) we'll find workarounds. I assume that the functionality would be available in qgis 2.99. Is that the case?

**Comment by signedav** [comment]
@saemiw Thanks. Yes this would be a qgis 2.99 fix that will be in 3.0.

@m-kuhn There would have been my other question. How is it with backwards compatibility? 
If I change to list, we wont be able to use the value map we configured in 2.18 when we open it with 3.0. Except I make a fallback if there was no list, I try to read the map. But this is code, mostly not used in future...

**Comment by m-kuhn** [comment]
there's a project file migration class (not sure the exact name) which can be used for this. or add the code in the config loader method of valemap drectly.

**Comment by signedav** [comment]
Fixed in Version 0.10.5 (in Playstore for Beta-Testers or [here](https://github.com/opengisch/QField/releases/tag/release-0.10.8_5)) with QGIS 2.99 (3.00)

**Comment by saemiw** [comment]
I can confirm this works nicely with QGIS 3 and QField 0.10.11. Thanks.

**Comment by saemiw** [comment]
I have to come back to this issue (which is QGIS not really Qfield-related). You did fix the actual issue. However when one imports values from a csv-file (which often happens if you have long lists of values that are repeatedly used) the import procedure orders the data according to the value (first column). The import procedure does not keep the order as it is shown in the csv file (when viewed with excel or notepadd++ for example). Do you see an option to also get the csv-import "right" (i.e. working the same way = keeping the entry order) in the QGIS-code?

**Comment by RovingGeologist** [comment]
I'm commenting because I would like to "second" saemiw's comment/request.  All of my value maps are populated with csv files that are exported from a database lookup table.  These are already sorted, but the sorting is changed during the important.

Depending on the field in the attribute table the value column can be either a text value or numeric value.  If it's a text value the value map can no longer be properly sorted.  For longer value map lists, or lists that are designed to have the most commonly used values at the top it will be a frustrating user experience.

In short: an second request for an option to keep the original order during csv-import into value maps.  I'll figure out how to submit this as a feature request for QGIS.

**Comment by saemiw** [comment]
Yes this would be nice. 

I work around it by importing the lists like that:
value, label
1, aaFirst Entry
2, Biever
3, zzAnother or last entry

Then I close and reopen the layer properties popup and remove the "aa" and "zz"

But its messy and it will be complicated if you have very long lists that have a complete random order. 

If you do have some money @RovingGeologist to fix this, we'd certainly help pay as well (but only in automn as we have used up our budget of the year already ;D).

**Comment by jarvojta** [comment]
This option is definitely needed.

Another work around is to edit the lists directly in qml files. Bulky as
well.
